### PR TITLE
[Hacktoberfest] Introducing autovalue

### PR DESCRIPTION
**Pull request by aballano**
- Introduce AutoValue and AutoValueMoshi as dependencies.
- Converted the Conversation class as an example (more to come later on).

Note: Please include the Hacktoberfest label :)

**Issue:** 
https://github.com/xing/xing-android-sdk/issues/137

Dependencies:__
- [x] Unit Tests

### CODE REVIEW

File: api-client/src/main/java/com/xing/api/XingApi.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
 package com.xing.api;
 
 import com.squareup.moshi.Moshi;
+import com.xing.api.internal.AutoValueFactory;
```
Comment (serj-lotutovici) [inline api-client/src/main/java/com/xing/api/XingApi.java]: `AutoValueFactory` should be in `com.xing.api.internal.json`

File: api-client/src/main/java/com/xing/api/data/messages/Conversation.java [hunk 2]
```diff
@@ -26,135 +29,43 @@
 /**
  * Java representation of a Conversation.
  */
-public class Conversation implements Serializable {
+@AutoValue
+public abstract class Conversation implements Serializable {
```
Comment (serj-lotutovici) [inline api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: Every class also needs a builder.

File: build.gradle [hunk 3]
```diff
@@ -79,6 +89,7 @@ allprojects {
 
 subprojects { project ->
 
+    apply plugin: 'net.ltgt.apt'
```
Comment (serj-lotutovici) [inline build.gradle]: Not all subprojects request apt. Only `:api-client` (the other one's are android libraries.

**Review by serj-lotutovici — CHANGES_REQUESTED** [review]
Knowing the XING api, this is a bit risky. AutoValue forces non nullability, which is a good thing, but not in our case.

Please consult the api documentation on dev.xing.com and make sure that all values are coming as non `null`.

Also please add a `@com.xing.api.internal.Nullable` annotation that can be used for fields that are potentially null.

File: api-client/src/main/java/com/xing/api/data/messages/Conversation.java (hunk 2 above)
Comment (aballano) [reply api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: I can add it, no problem although I don´t see the need since those models are only intended to be created through moshi, right?

Comment (serj-lotutovici) [reply api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: try to write a test where one field is not present or null. It should crash with an NPE since AutoValues adds validations in the constructer of the generated class.

Comment (serj-lotutovici) [reply api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: Strike that... in the email looked like, you where asking about the nullability.

Comment (serj-lotutovici) [reply api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: You're probably right... this class may not need a builder.

Comment (aballano) [reply api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: Is it really necessary? Think that in order to do it properly we would need 3 tests per object field: one with normal value -> no exception and two more for null-not present value -> exception. That means making 8*3 = 24 tests just to check that autovalue throws an exception in that case... I don't see the point.

Going back to the builder: If we're gonna create this objects manually I can add it, otherwise if it's gonna be only created by moshi (which is the case I think) I don't think is necessary.

**Comment by aballano** [comment]
Checked the api and the only field which could potentially be null is the `latestMessages`, so I added the annotation there.

File: api-client/src/main/java/com/xing/api/data/messages/Conversation.java (hunk 2 above)
Comment (serj-lotutovici) [reply api-client/src/main/java/com/xing/api/data/messages/Conversation.java:34]: > Going back to the builder: If we're gonna create this objects manually I can add it, otherwise if it's gonna be only created by moshi (which is the case I think) I don't think is necessary.

In any case, we can add the builder later if we need it. So ignored for now.

> Is it really necessary? Think that in order to do it properly we would need 3 tests per object field: one with normal value -> no exception and two more for null-not present value -> exception. That means making 8*3 = 24 tests just to check that autovalue throws an exception in that case... I don't see the point.

The point is not to add tests for each case. The point is to make sure, that the api will always return non null values. Otherwise, the possible nulls should be annotated, otherwise we break the usage of the api.

File: api-client/src/main/java/com/xing/api/internal/Nullable.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Informative annotation that describes that a parameter or a method is experimental.
+ *
+ * @author serj.lotutovici
+ */
+@Documented
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.CLASS)
+public @interface Experimental {
```
Comment (serj-lotutovici) [inline api-client/src/main/java/com/xing/api/internal/Nullable.java]: Should be `Nullable`

File: api-client/src/main/java/com/xing/api/internal/Nullable.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Informative annotation that describes that a parameter or a method is experimental.
+ *
+ * @author serj.lotutovici
```
Comment (serj-lotutovici) [inline api-client/src/main/java/com/xing/api/internal/Nullable.java]: no need for author

File: api-client/src/main/java/com/xing/api/internal/Nullable.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+package com.xing.api.internal;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Informative annotation that describes that a parameter or a method is experimental.
```
Comment (serj-lotutovici) [inline api-client/src/main/java/com/xing/api/internal/Nullable.java]: Also wrong doc

Comment (aballano) [reply api-client/src/main/java/com/xing/api/internal/Nullable.java]: Yeah, forgot to push this entire change actually

**Review by serj-lotutovici — APPROVED** [review]

**Comment by serj-lotutovici** [comment]
Thanks!
